packages/core/src/network/heartbeat.ts

```diff
@@ -38,69 +63,113 @@ export default class Heartbeat {
 
   public handleHeartbeatRequest(msg: Uint8Array, remotePeer: PeerId): Promise<Uint8Array> {
     this.networkPeers.register(remotePeer)
-    log('beat')
+    log(`received heartbeat from ${remotePeer.toB58String()}`)
     return Promise.resolve(Heartbeat.calculatePingResponse(msg))
   }
 
-  public async pingNode(id: PeerId): Promise<boolean> {
-    log('ping', id.toB58String())
+  /**
+   * Attempts to ping another peer.
+   * @param destination id of the node to ping
+   * @param signal [optional] abort controller to prematurely end request
+   * @returns a Promise of a pingResult object with property `lastSeen < 0` if there were a timeout
+   */
+  public async pingNode(destination: PeerId, signal?: AbortSignal): Promise<HeartbeatPingResult> {
+    log('ping', destination.toB58String())
 
     const challenge = randomBytes(16)
     const expectedResponse = Heartbeat.calculatePingResponse(challenge)
 
+    let pingResponse: Uint8Array[] | undefined
+
     try {
-      const pingResponse = await this.sendMessage(id, this.protocolHeartbeat, challenge, true, {
-        timeout: HEARTBEAT_TIMEOUT
+      pingResponse = await this.sendMessage(destination, this.protocolHeartbeat, challenge, true, {
+        timeout: this.config.heartbeatDialTimeout,
+        signal
       })
+    } catch (err) {
+      log(`Connection to ${destination.toB58String()} failed: ${err?.message}`)
+      return {
+        destination,
+        lastSeen: -1
+      }
+    }
+
+    if (pingResponse == null || pingResponse.length != 1 || !u8aEquals(expectedResponse, pingResponse[0])) {
+      log(`Mismatched challenge. Got ${u8aToHex(pingResponse[0])} but expected ${u8aToHex(expectedResponse)}`)
+      await this.hangUp(destination)
 
-      if (pingResponse == null || pingResponse.length == 0 || !u8aEquals(expectedResponse, pingResponse[0])) {
-        log(`Mismatched challenge. ${pingResponse}`)
-        await this.hangUp(id)
-        return false
+      return {
+        destination,
+        lastSeen: -1
       }
+    }
 
-      log('ping success to', id.toB58String())
-      return true
-    } catch (e) {
-      log(`Connection to ${id.toB58String()} failed: ${JSON.stringify(e)}`)
-      return false
+    return {
+      destination,
+      lastSeen: Date.now()
     }
   }
 
-  private async checkNodes(): Promise<void> {
-    const thresholdTime = Date.now() - HEARTBEAT_INTERVAL
+  /**
+   * Performs a ping request to all nodes who were not seen since the threshold
+   */
+  protected async checkNodes(): Promise<void> {
+    const thresholdTime = Date.now() - this.config.heartbeatThreshold
     log(`Checking nodes since ${thresholdTime} (${new Date(thresholdTime).toLocaleString()})`)
 
-    const toPing = this.networkPeers.pingSince(thresholdTime)
+    const abort = new AbortController()
+
+    let finished = false
+
+    setTimeout(() => {
```

Comment:
Why adding yet another timeout for aborting the previous tries?

Reply:
The idea is to make sure that `nAtATime` will terminate no matter what happens. Since finished is set to true once `nAtATime` comes back there won't be any abort afterwards.